src/transaction/requests.rs

```diff
@@ -379,35 +380,51 @@ impl Merge<ResponseWithShard<kvrpcpb::PessimisticLockResponse, Vec<kvrpcpb::Muta
             Result<ResponseWithShard<kvrpcpb::PessimisticLockResponse, Vec<kvrpcpb::Mutation>>>,
         >,
     ) -> Result<Self::Out> {
-        input
-            .into_iter()
-            .flat_map_ok(|ResponseWithShard(mut resp, mutations)| {
-                let values = resp.take_values();
-                let values_len = values.len();
-                let not_founds = resp.take_not_founds();
-                let kvpairs = mutations
-                    .into_iter()
-                    .map(|m| m.key)
-                    .zip(values)
-                    .map(KvPair::from);
-                assert_eq!(kvpairs.len(), values_len);
-                if not_founds.is_empty() {
-                    // Legacy TiKV does not distiguish not existing key and existing key
-                    // that with empty value. We assume that key does not exist if value
-                    // is empty.
-                    Either::Left(kvpairs.filter(|kvpair| !kvpair.value().is_empty()))
-                } else {
-                    assert_eq!(kvpairs.len(), not_founds.len());
-                    Either::Right(kvpairs.zip(not_founds).filter_map(|(kvpair, not_found)| {
-                        if not_found {
-                            None
-                        } else {
-                            Some(kvpair)
-                        }
-                    }))
-                }
+        let (success, mut errors): (Vec<_>, Vec<_>) = input.into_iter().partition(Result::is_ok);
```

Comment:
It's unncessary to always partition which introduces overhead of memory copy. We can check if there is error first.